File: apps/flutter/lib/features/categories/domain/entities/category.dart
class Category {
  final String id;
  final String name;
  final String kind;
  final String? parentId;
  final String? color;
  final String? icon;
  final bool isActive;

  const Category({
    required this.id,
    required this.name,
    required this.kind,
    required this.parentId,
    required this.color,
    required this.icon,
    required this.isActive,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: (json["categoryId"] ?? json["id"]) as String,
      name: json["name"] as String,
      kind: json["kind"] as String,
      parentId: json["parentId"] as String?,
      color: json["color"] as String?,
      icon: json["icon"] as String?,
      isActive: json["isActive"] as bool? ?? true,
    );
  }
}

File: apps/api/src/models/category.ts
import { AggregateRoot } from "@abejarano/ts-mongodb-criteria"
import { createMongoId } from "./shared/mongo_id"

export type CategoryKind = "income" | "expense"

export type CategoryPrimitives = {
  id?: string
  categoryId: string
  userId: string
  name: string
  kind: CategoryKind
  parentId?: string | null
  color?: string | null
  icon?: string | null
  isActive: boolean
  createdAt: Date
  updatedAt?: Date
}

export type CategoryCreateProps = {
  userId: string
  name: string
  kind: CategoryKind
  parentId?: string | null
  color?: string | null
  icon?: string | null
  isActive?: boolean
}

export class Category extends AggregateRoot {
  private readonly props: CategoryPrimitives

  private constructor(props: CategoryPrimitives) {
    super()
    this.props = props
  }

  static create(props: CategoryCreateProps): Category {
    const now = new Date()

    return new Category({
      categoryId: createMongoId(),
      userId: props.userId,
      name: props.name,
      kind: props.kind,
      parentId: props.parentId ?? null,
      color: props.color ?? null,
      icon: props.icon ?? null,
      isActive: props.isActive ?? true,
      createdAt: now,
      updatedAt: now,
    })
  }

  getId(): string {
    return this.props.id ?? this.props.categoryId
  }

  getCategoryId(): string {
    return this.props.categoryId
  }

  toPrimitives(): CategoryPrimitives {
    return this.props
  }

  static override fromPrimitives(primitives: CategoryPrimitives): Category {
    return new Category(primitives)
  }
}
